### PR TITLE
Adding Architectural Pattern doc

### PR DESCRIPTION
**Pull request by UkiahSmith**
This document is ready to be published. There are graphics forthcoming.

### CODE REVIEW

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+An important part of setting up a new Kong cluster is choosing the right architectural pattern that fits your needs and integrates with your existing infrastructure. Here are a few that we commonly see and recommend. Kong itself is greatly flexible, and can be setup in many ways. We present these as starting points in hopes that you find that using or extending one of these patterns provides an ideal fit for your unique needs.
+
+## Kong between client and upstream
+
+The classic architectural pattern we see customers setting up is adding Kong in between some, or all, of their API traffic. Kong is installed in your prefered hosting provider, and configured to proxy traffic to the existing APIs. Note that in this pattern the APIs are still accessible directly, that is both through the proxy and also without being routed through the API Gateway. We see this during initial testing, and migration phases, before customers eventually route all traffic through Kong. 
+
+The pros of this approach are that it is simple to setup and configure. All it requires is setup, and coordinating new traffic to use the gateway. Older traffic will still flow directly to the APIs.
```
Comment (travispamaral) [inline app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: I would break Pros & Cons into `<h3>` tags to help make them stand out.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+## Multi-datacenter with Cassandra
+
+Application availability is important, and high availability can be achieved by configuring Kong in a multi-datacenter setup using [Apache Cassandra](https://getkong.org/about/faq/#apache-cassandra) as the backing datastore. In this configuration each geographically disparate datacenter is host to at least three Kong nodes, and the appropriate number of Cassandra nodes.
+
+The minimum number of Kong nodes per data center needed for high availability is three. This will ensure that in the unlikely event of a node failure there will still be a node and a backup. Cassandra would be configured in an optimal way; you can use the [Cassandra Calculator](https://www.ecyrd.com/cassandracalculator/) to determine how many Cassandra nodes are needed per data center to ensure high availability, and  performance for your specific needs.
```
Comment (travispamaral) [inline app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: I would reword the sentence "Casandra would be configured in an optimal way;"  Maybe something like "_To ensure Casandra is configured int the an optimal way we recommend using the [Casandra Calculator]_"

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+Each node in Kong cluster can be configured using its kong.conf file. Two separate configurations can be created, one for a proxy only node, and the other for an admin only node. The proxy only nodes would be accessible from the outside world. They would connect to the database and the upstream APIs that are behind the firewall. The admin only node would be inside the firewall, or a separate firewall (or VPC on AWS), and it would connect to the database that the proxy nodes connects to.
+
+So long as the proxy nodes and admin nodes use the same database they will be part of the same cluster. The admin node itself can be located anywhere. There are a few different ways to achieve this. read [securing the Kong Admin API](https://getkong.org/docs/latest/secure-admin-api/) for details.
+
+### Unlimited options
+
+Kong as a tool is flexible, and offers many possibilities for incorporating it into your existing infrastructure. These patterns we touch upon are the ones we commonly see and recommend. Every situation is unique, and we are happy to assist you with your specific needs.
```
Comment (darrenjennings) [inline app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: Suggestion: add in a way for someone to contact us. Not sure what is consistent, link to contact form, email, phone number... etc.

File: app/_data/docs_nav_ee_0.31-x.yml [hunk 4]
```diff
@@ -64,13 +64,15 @@
...
   items:
-    - text: Kong Implementation Checklist
-      url: /kong-implementation-checklist
-    - text: Changelog
+    - text: Kong Enterprise Changelog
       url: /docs/enterprise/changelog
       absolute_url: true
     - text: Architecture Overview
       url: /kong-architecture-overview
+    - text: Architecture Patterns
```
Comment (darrenjennings) [inline app/_data/docs_nav_ee_0.31-x.yml]: would recommend nesting into Architecture and then having two sections: overview and patterns. But would advise keeping the url for overview since it's been a page so long and should keep SEO.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+
+External Links
```
Comment (darrenjennings) [inline app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: seems strange to call this section External links since all but one is on the same domain as the document the reader is viewing. Maybe "Reference Links"? Github link should point to https://github.com/kong/kong, and would change "Kong Documentation" to "Kong CE Documentation" to make it obvious they're ejecting from the enterprise documentation. Also `/install` is actually just a redirect to https://konghq.com/install/ so maybe we can link directly here.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md (hunk 1 above)
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: Noted.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md (hunk 2 above)
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: Noted.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
+---
+title: Kong Implementation Checklist
+---
+# Kong Architecture Patterns
+
+An important part of setting up a new Kong cluster is choosing the right architectural pattern that fits your needs and integrates with your existing infrastructure. Here are a few that we commonly see and recommend. Kong itself is greatly flexible, and can be setup in many ways. We present these as starting points in hopes that you find that using or extending one of these patterns provides an ideal fit for your unique needs.
```
[the remark this answers is not in the document]
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: Noted.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+## Kong and APIs behind firewall
+
+The next pattern we see, and the next step for security, is to setup [API servers behind a firewall](https://getkong.org/docs/latest/network/) and only allow the Kong API Gateway be accessible to the public.
+
+This is a popular architecture pattern that we see with customers who have publicly available APIs. Either they are migrating from the initial pattern, improving security by cordoning off direct access to the APIs, or choosing it for their general use. 
+
+This architecture pattern is the one we recommend for most customers. It is straightforward to setup, and offers security and control over API access.
+
+The Kong cluster would be placed behind a firewall, or within an Amazon VPC. Port 80 would be opened to the outside world for just the Kong cluster. Inside the firewall the API servers would only have allowed connections from the Kong cluster. Other IP:ports may need to be opened for the API servers to reach resources they need to fulfill any requests.
```
[the remark this answers is not in the document]
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: Tweaked.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+
+Access to the [Kong Admin API can be secured in a couple of different ways](https://getkong.org/docs/latest/secure-admin-api/). Depending on your needs you would lock it down to localhost on the machine it is running on, use the firewall to prevent access from anyone outside your network, or proxy the Admin API through Kong itself, and secure it using one of the authentication plugins.
+
+With Kong Enterprise you also have the option of using Role Based Access Control (RBAC) to shape fine grained access for your team, and co-workers. [RBAC in Kong](https://getkong.org/docs/enterprise/0.31-x/plugins/rbac-api/) is used to create roles with defined permissions; roles are then assigned to users who are limited in what they can and can not do by the permissions. Roles could be created for read-only roles; for when you want to let co-workers look, but not edit, the configuration. You could also create roles for different teams; giving them access to edit and update only their APIs.
```
[the remark this answers is not in the document]
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: 👍

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+## Separate Kong Admin node
+
+For security conscious customers we recommend a pattern of having a single node in a Kong cluster allow access to the Admin API. 
+
+Each node in Kong cluster can be configured using its kong.conf file. Two separate configurations can be created, one for a proxy only node, and the other for an admin only node. The proxy only nodes would be accessible from the outside world. They would connect to the database and the upstream APIs that are behind the firewall. The admin only node would be inside the firewall, or a separate firewall (or VPC on AWS), and it would connect to the database that the proxy nodes connects to.
+
+So long as the proxy nodes and admin nodes use the same database they will be part of the same cluster. The admin node itself can be located anywhere. There are a few different ways to achieve this. read [securing the Kong Admin API](https://getkong.org/docs/latest/secure-admin-api/) for details.
```
[the remark this answers is not in the document]
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: 👍

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md (hunk 3 above)
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: I don't care for this. No other docs have this. Audience should be existing customers.

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md (hunk 5 above)
Comment (UkiahSmith) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: 👍

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md (hunk 2 above)
Comment (Tieske) [reply app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: "and  performance" has 2 spaces

File: app/docs/enterprise/0.31-x/kong-architecture-patterns.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+## Multi-datacenter with Cassandra
+
+Application availability is important, and high availability can be achieved by configuring Kong in a multi-datacenter setup using [Apache Cassandra](https://getkong.org/about/faq/#apache-cassandra) as the backing datastore. In this configuration each geographically disparate datacenter is host to at least three Kong nodes, and the appropriate number of Cassandra nodes.
+
+The minimum number of Kong nodes per data center needed for high availability is three. This will ensure that in the unlikely event of a node failure there will still be a node and a backup. Cassandra would be configured in an optimal way; you can use the [Cassandra Calculator](https://www.ecyrd.com/cassandracalculator/) to determine how many Cassandra nodes are needed per data center to ensure high availability, and  performance for your specific needs.
+
+Logically Kong will act as a single cluster, and in a multi-datacenter strategy, can be configured similar to the previous architecture patterns. We recommending having the upstream services behind a firewall (or in a VPC in AWS) and only allow Kong nodes direct access to the APIs.
```
Comment (Tieske) [inline app/docs/enterprise/0.31-x/kong-architecture-patterns.md]: we recommending -> we recommend
